File: Editorial/109_Convert_Sorted_List_to_Binary_Search_Tree/Go/3_Inorder_Simulation.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 *
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
var head *ListNode

func findSize(head *ListNode) int {
    ptr := head
    c := 0
    for ptr != nil {
        ptr = ptr.Next
        c += 1
    }
    return c
}

func convertListToBST(l int, r int) *TreeNode {
    if l > r {
        return nil
    }
    mid := (l + r) / 2
    left := convertListToBST(l, mid-1)
    var node *TreeNode
    if head != nil {
        node = &TreeNode{Val: head.Val}
        node.Left = left
        head = head.Next
        node.Right = convertListToBST(mid+1, r)
    }
    return node
}

func sortedListToBST(headParam *ListNode) *TreeNode {
    size := findSize(headParam)
    head = headParam
    return convertListToBST(0, size-1)
}

File: Editorial/109_Convert_Sorted_List_to_Binary_Search_Tree/JavaScript/3_Inorder_Simulation.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
var sortedListToBST = function (head) {
    var findSize = function (head) {
        let ptr = head,
            c = 0;
        while (ptr !== null) {
            ptr = ptr.next;
            c += 1;
        }
        return c;
    };
    var convertListToBST = function (l, r) {
        if (l > r) return null;
        let mid = Math.floor((l + r) / 2);
        let left = convertListToBST(l, mid - 1);
        let node = new TreeNode(head.val);
        node.left = left;
        head = head.next;
        node.right = convertListToBST(mid + 1, r);
        return node;
    };
    let size = findSize(head);
    return convertListToBST(0, size - 1);
};
